source/opt/constant_value.h

```diff
@@ -0,0 +1,321 @@
+
+#include "types.h"
+
+namespace spvtools {
+namespace opt {
+
+// Class hierarchy to represent the normal constants defined through
+// OpConstantTrue, OpConstantFalse, OpConstant, OpConstantNull and
+// OpConstantComposite instrcutions.
+class Const;
+class ScalarConst;
+class IntConst;
+class FloatConst;
+class BoolConst;
+class CompositeConst;
+class StructConst;
+class VectorConst;
+class ArrayConst;
+class NullConst;
+
+// Abstract class for a SPIR-V constant. It has a bunch of As<subclass> methods,
+// which is used as a way to probe the actual <subclass>
+class Const {
+ public:
+  Const() = delete;
+  virtual ~Const() {}
+
+  // Make a copy of this constant.
+  virtual std::unique_ptr<Const> Copy() const = 0;
+
+  // reflections
+  virtual ScalarConst* AsScalarConst() { return nullptr; }
+  virtual IntConst* AsIntConst() { return nullptr; }
+  virtual FloatConst* AsFloatConst() { return nullptr; }
+  virtual BoolConst* AsBoolConst() { return nullptr; }
+  virtual CompositeConst* AsCompositeConst() { return nullptr; }
+  virtual StructConst* AsStructConst() { return nullptr; }
+  virtual VectorConst* AsVectorConst() { return nullptr; }
+  virtual ArrayConst* AsArrayConst() { return nullptr; }
+  virtual NullConst* AsNullConst() { return nullptr; }
+
+  virtual const ScalarConst* AsScalarConst() const { return nullptr; }
+  virtual const IntConst* AsIntConst() const { return nullptr; }
+  virtual const FloatConst* AsFloatConst() const { return nullptr; }
+  virtual const BoolConst* AsBoolConst() const { return nullptr; }
+  virtual const CompositeConst* AsCompositeConst() const { return nullptr; }
+  virtual const StructConst* AsStructConst() const { return nullptr; }
+  virtual const VectorConst* AsVectorConst() const { return nullptr; }
+  virtual const ArrayConst* AsArrayConst() const { return nullptr; }
+  virtual const NullConst* AsNullConst() const { return nullptr; }
+
+  const analysis::Type* type() const { return type_; }
+
+ protected:
+  Const(const analysis::Type* ty) : type_(ty) {}
+
+  // The type of this constant.
+  const analysis::Type* type_;
+};
+
+// Abstract class for scalar type constants.
+class ScalarConst : public Const {
+ public:
+  ScalarConst() = delete;
+  ScalarConst* AsScalarConst() override { return this; }
+  const ScalarConst* AsScalarConst() const override { return this; }
+
+  // Returns a copy of the value of this constant in 32-bit words.
+  virtual std::vector<uint32_t> words() const = 0;
+
+ protected:
+  ScalarConst(const analysis::Type* ty) : Const(ty) {}
+};
+
+// Integer type constant.
+class IntConst : public ScalarConst {
+ public:
+  IntConst(const analysis::Type* ty, const std::vector<uint32_t>& words)
+      : ScalarConst(ty), words_(words) {}
+
+  IntConst* AsIntConst() override { return this; }
+  const IntConst* AsIntConst() const override { return this; }
+
+  std::unique_ptr<IntConst> CopyIntConst() const {
+    return std::unique_ptr<IntConst>(new IntConst(type_, words_));
```

Comment:
I've submitted f51d82338e6a129300562b2c73daee02d3918b77, which can be a little helpful to make this shorter. ;)